pkg/cache/resource_test.go

```diff
@@ -0,0 +1,275 @@
+		},
+		"cqs with cohort": {
+			clusterQueues: []kueue.ClusterQueue{
+				utiltesting.MakeClusterQueue("cq1").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).ClusterQueue,
+				utiltesting.MakeClusterQueue("cq2").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10", "", "9").Obj(),
+					).ClusterQueue,
+			},
+			cohorts: []kueuealpha.Cohort{
+				utiltesting.MakeCohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).Cohort,
+			},
+			usage: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 1000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 500},
+			},
+			wantAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 28_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 28_500},
+			},
+			wantPotentiallyAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 29_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 30_000},
+			},
+		},
+		"cq borrows from cohort": {
+			clusterQueues: []kueue.ClusterQueue{
+				utiltesting.MakeClusterQueue("cq1").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).ClusterQueue,
+			},
+			cohorts: []kueuealpha.Cohort{
+				utiltesting.MakeCohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).Cohort,
+			},
+			usage:                    map[string]resources.FlavorResourceQuantities{"cq1": {{Flavor: "red", Resource: "cpu"}: 11000}},
+			wantAvailable:            map[string]resources.FlavorResourceQuantities{"cq1": {{Flavor: "red", Resource: "cpu"}: 9_000}},
+			wantPotentiallyAvailable: map[string]resources.FlavorResourceQuantities{"cq1": {{Flavor: "red", Resource: "cpu"}: 20_000}},
+		},
+		"cq oversubscription spills into cohort": {
+			clusterQueues: []kueue.ClusterQueue{
+				utiltesting.MakeClusterQueue("cq1").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).ClusterQueue,
+				utiltesting.MakeClusterQueue("cq2").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10", "0", "0").Obj(),
+					).ClusterQueue,
+			},
+			cohorts: []kueuealpha.Cohort{
+				utiltesting.MakeCohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).Cohort,
+			},
+			usage: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 31_000},
+			},
+			wantAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: -11000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: -1000},
```

Comment:
interesting....
What would happen if those 10 CPUs that CQ2 has are split into two CQs?
What would the usage on each of them be, and why?

Reply:
`cohort` SubtreeQuota is 20k (from `cohort`, and CQ1). Usage in cohort is 31k.  When checking usage, CQ2 and CQ3 would find 5k available locally and -11k in the parent. Adding those numbers would result in `available` -6k for each of those CQs.

Reply:
When a negative number is returned, it indicates that some CQ/Cohort in the CohortTree is using too much resources - which can happen after a change of available resources, or switching parent Cohort.

This ensures that, for any CohortTree, we won't admit a workload that would result in using more of a FlavorResource than is in the CohortTree. I think it is desirable that this also affects CQs which do not lend capacity to or borrow capacity from the CohortTree - once we are overadmited, we should prevent any admissions which worsen the debt.

Reply:
As discussed offline, consider just returning zero when the CQ is oversubscribed.
IIUC, it shouldn't have any effect on algorithms, and this number is not reported to the user.
The advantage is just more maintainable unit tests.

Reply:
Done. We report the capacity lacked to the user during flavor assignment. Updated the overadmission integration test with a new number.